### PR TITLE
Add KKS port

### PR DESCRIPTION
**Pull request by ManlyMarco**
Appears to be working correctly.

### CODE REVIEW

**Comment by ManlyMarco** [comment]
Confirmed to be working by people on discord.

**Comment by MayouKurayami** [comment]
> Confirmed to be working by people on discord.

Has it been tested in VR (both official and mod)?
I don't own or play KKS so I can't test it myself.

File: Core_BetterSquirt/CharaSquirt.cs [hunk 1]
```diff
@@ -157,7 +171,7 @@ private bool CheckTwitchSECond()
 
 		private static bool PlaySetting(Utils.Sound.Setting setting, Transform referenceInfo)
 		{
-			Transform soundsource = Utils.Sound.Play(setting);
+			Transform soundsource = Utils.Sound.Play(setting).transform;
```
Comment (MayouKurayami) [inline Core_BetterSquirt/CharaSquirt.cs:171]: Is this change required for the port?
If not, would you mind separating it into a different commit and explain the change?

File: KK_BetterSquirt/packages.config
Comment (MayouKurayami) [inline KK_BetterSquirt/packages.config:1]: Is there any particular reason why the dependencies are updated for KK?
I'd prefer to leave the dependencies alone on the KK side unless it's necessary.

File: Core_BetterSquirt/CharaSquirt.cs [hunk 2]
```diff
@@ -33,7 +33,13 @@ internal float TouchCooldown
 		}
 
 		private static readonly MethodInfo hitReactionPlayInfo = AccessTools.Method(
-			Type.GetType("VRHandCtrl, Assembly-CSharp") ?? typeof(HandCtrl), "HitReactionPlay", new Type[] { typeof(int), typeof(bool) });
+			Type.GetType("VRHandCtrl, Assembly-CSharp") ?? typeof(HandCtrl), 
+            nameof(HandCtrl.HitReactionPlay), 
+#if KK
+            new Type[] { typeof(int), typeof(bool) });
+#else
+            new Type[] { typeof(AibuColliderKind), typeof(bool) });
```
Comment (MayouKurayami) [inline Core_BetterSquirt/CharaSquirt.cs]: Is this change necessary?
The `AibuColliderKind` type referenced here is from the `HandCtrl` class, which doesn't match the signature of `VRHandCtrl.HitReactionPlay()`.

I'd prefer to not needlessly break compatibility with the official VR module unless there is a very good reason to.

**Comment by ManlyMarco** [comment]
It works in KKS_VR but not in the official module (nothing works in it anyways so it doesn't matter).

**Review by MayouKurayami — CHANGES_REQUESTED** [review]

File: Core_BetterSquirt/CharaSquirt.cs (hunk 1 above)
Comment (ManlyMarco) [reply Core_BetterSquirt/CharaSquirt.cs:171]: It's required in KKS.

File: Core_BetterSquirt/CharaSquirt.cs (hunk 2 above)
Comment (ManlyMarco) [reply Core_BetterSquirt/CharaSquirt.cs]: This is necessary for KKS since the method signature changed. There is no change for KK.

File: KK_BetterSquirt/packages.config
Comment (ManlyMarco) [reply KK_BetterSquirt/packages.config:1]: OnStartH overload with MonoBehaviour parameter needs a more recent version of KKAPI than what is currently used.

File: Core_BetterSquirt/CharaSquirt.cs (hunk 2 above)
Comment (MayouKurayami) [reply Core_BetterSquirt/CharaSquirt.cs]: I just checked the decompiled code of KKS. 
The signature for both KK and KKS is the same `HitReactionPlay(AibuColliderKind, bool)`, where `AibuColliderKind` is an enum defined in either `HandCtrl` or `VRHandCtrl`.

The reason I use `int` here is so that it would work with both `HandCtrl.AibuColliderKind` and `VRHandCtrl.AibuColliderKind`. This works fine in KK.
Does using `int` not work in KKS?

File: KK_BetterSquirt/packages.config
Comment (MayouKurayami) [reply KK_BetterSquirt/packages.config:1]: Use a preprocessor directive and only use the OnStartH overload with MonoBehaviour parameter for KKS?

File: Core_BetterSquirt/CharaSquirt.cs (hunk 2 above)
Comment (ManlyMarco) [reply Core_BetterSquirt/CharaSquirt.cs]: Using int as a substitute for an enum with underlying int for GetMethod and other reflection only works in .Net 3.5 I believe. It's not recommended either way, the proper type should be used instead.

Comment (ManlyMarco) [reply Core_BetterSquirt/CharaSquirt.cs]: It can be done instead with
```cs
		private static readonly MethodInfo hitReactionPlayInfo = 
			Type.GetType("VRHandCtrl, Assembly-CSharp")?.GetMethod(nameof(HandCtrl.HitReactionPlay), AccessTools.all, null, new[] { typeof(int), typeof(bool) }, null) ?? 
			typeof(HandCtrl).GetMethod(nameof(HandCtrl.HitReactionPlay), AccessTools.all, null, new[] { typeof(HandCtrl.AibuColliderKind), typeof(bool) }, null) ;
```

File: KK_BetterSquirt/packages.config
Comment (ManlyMarco) [reply KK_BetterSquirt/packages.config:1]: There is little point in doing so when the proper way is to use the MonoBehaviour overload and have identical code for both games.

Comment (MayouKurayami) [reply KK_BetterSquirt/packages.config:1]: The point is to avoid all the other dependencies that are required by KKAPI newer than v1.23, such as the modified Assembly-CSharp nuget package and the overhauled MoreAccessories, with all the potential instability and incompatibility they may introduce. 
 
The KK side of the plugin works fine as it is, and I don't want to jeopardize its stability and backward compatibility for the sake of KKS.

File: Core_BetterSquirt/CharaSquirt.cs (hunk 2 above)
Comment (MayouKurayami) [reply Core_BetterSquirt/CharaSquirt.cs]: > Using int as a substitute for an enum with underlying int for GetMethod and other reflection only works in .Net 3.5 I believe.

Have you tested the original code in KKS to confirm that it doesn't work?

*******

```c#
Type.GetType("VRHandCtrl, Assembly-CSharp")?.GetMethod(nameof(HandCtrl.HitReactionPlay), AccessTools.all, null, new[] { typeof(int), typeof(bool) }, null)
```
Will this work in KKS official VR?  Or is it intended to simply return null in that case?

I'd prefer to not break compatibility with the KKS official VR module unless this plugin is confirmed to not work at all, in which case the rest of the VR related code should be removed on the KKS side.

Comment (ManlyMarco) [reply Core_BetterSquirt/CharaSquirt.cs]: > Have you tested the original code in KKS to confirm that it doesn't work?

Yes, it doesn't work. Otherwise I wouldn't have changed it.

I have not tested it with the official VR module and it won't run in it anyways since the process filter is missing. It might work in official KKS VR as it is for all I know.

Even if it does not work in the KKS VR module, it doesn't make sense to increase code complexity by removing the VR code in KKS since then there's more unique code paths that can have their own bugs and need additional testing.

File: KK_BetterSquirt/packages.config
Comment (ManlyMarco) [reply KK_BetterSquirt/packages.config:1]: I don't share that view, the things you've mentioned have been in use for a long time by now and have proven to be stable, but fair enough.

File: Core_BetterSquirt/CharaSquirt.cs (hunk 2 above)
Comment (ManlyMarco) [reply Core_BetterSquirt/CharaSquirt.cs]: Just tested the KKS VR module and the plugin does not work at all. Many classes are different from the main game as well as the KK VR module. Making it work would require changing a lot of the code into pure reflection calls.

Comment (MayouKurayami) [reply Core_BetterSquirt/CharaSquirt.cs]: > Even if it does not work in the KKS VR module, it doesn't make sense to increase code complexity by removing the VR code in KKS since then there's more unique code paths that can have their own bugs and need additional testing.

Right, I forgot that you can just use the process filter to prevent it from causing issues in KKS VR.

File: KK_BetterSquirt/packages.config
Comment (ManlyMarco) [reply KK_BetterSquirt/packages.config:1]: I've reverted the KKAPI package, is it good enough?

File: Core_BetterSquirt/CharaSquirt.cs (hunk 2 above)
Comment (ManlyMarco) [reply Core_BetterSquirt/CharaSquirt.cs]: Yes, the plugin only has the main game process filter so it does not load in the VR module.

**Review by MayouKurayami — APPROVED** [review]

**Comment by MayouKurayami** [comment]
If everything looks ok and is tested to be working in KKS in its current state then I'll proceed with the merge

**Review by ManlyMarco — COMMENTED** [review]
Everything looks fine, should be safe to merge and make a release.
